### PR TITLE
Taxonomy benchmark

### PR DESCRIPTION
**Pull request by sgalkina**
A command to benchmark taxonomic annotations in the absense of ground truth used for Figure 3 scatter plots.

### CODE REVIEW

File: vamb/__main__.py [hunk 1]
```diff
@@ -1559,6 +1565,244 @@ def predict_taxonomy(
...
+    tnfs: np.ndarray,
+    lengths: np.ndarray,
+    out_dir: Path,
+    taxonomy_options: TaxometerOptions,
+    cuda: bool,
+) -> vamb.taxonomy.PredictedTaxonomy:
+    logger.info("Running cross validation for the taxonomy")
+
+    tax = taxonomy_options.taxonomy
+    t = vamb.taxonomy.Taxonomy.from_file(tax.path, comp_metadata, False)
```
Comment (jakobnissen) [inline vamb/__main__.py:1678]:
```suggestion
    taxonomy = vamb.taxonomy.Taxonomy.from_file(tax.path, comp_metadata, False)
```

File: vamb/__main__.py [hunk 2]
```diff
@@ -1559,6 +1565,244 @@ def predict_taxonomy(
...
+    for i in t.contig_taxonomies:
+        if i is None or len(i.ranks) == 0:
+            classes_order.append("root")
+        else:
+            classes_order.append(i.ranks[-1])
+    targets = np.array([ind_nodes[i] for i in classes_order])
+
+    def train_and_predict_fold(fold, train_idx, test_idx):
+        logger.info(
+            f"Fold {fold + 1}/{k}: Training on {len(train_idx)} contigs, testing on {len(test_idx)} contigs"
```
Comment (jakobnissen) [inline vamb/__main__.py:1700]: I'm not a big fan of nested functions - they break up the original function, and the way they read variables from the surrounding scope can be confusing. Move to global scope.

File: vamb/__main__.py [hunk 3]
```diff
@@ -1559,6 +1565,244 @@ def predict_taxonomy(
...
+    tax = taxonomy_options.taxonomy
+    t = vamb.taxonomy.Taxonomy.from_file(tax.path, comp_metadata, False)
+    n_contigs = len(t.contig_taxonomies)
+    orig_path = str(tax.path)
+    predicted_path = out_dir.joinpath(f"results_taxonomy_predicted_kfold.tsv")
+    file_tracking = out_dir.joinpath(f"file_tracking.tsv")
+    accuracy_file = out_dir.joinpath(f"accuracy_report.tsv")
+
+    all_predicted_taxonomies = []
+    k = 5
```
Comment (jakobnissen) [inline vamb/__main__.py:1686]: Maybe `k` could be a user-specified parameter defaulting to 5? Not a big deal.

File: vamb/__main__.py [hunk 4]
```diff
@@ -1559,6 +1565,244 @@ def predict_taxonomy(
...
+    fold_args = [
+        (fold, train_idx, test_idx)
+        for fold, (train_idx, test_idx) in enumerate(kf.split(np.arange(n_contigs)))
+    ]
+    loss_tests = []
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        results = list(
+            executor.map(lambda args: train_and_predict_fold(*args), fold_args)
+        )
```
Comment (jakobnissen) [inline vamb/__main__.py:1787]: I'm not sure this is a good idea. The bottleneck is presumably the neural network training, which is already multithreaded. Running multiple trainings on the same CPU will oversubscribe threads, so it could potentially be less efficient than just running it in serial.
Have you checked that running concurrently is faster? Perhaps the network training can run in serial, and then the benchmarking can run concurrently?

File: vamb/__main__.py [hunk 5]
```diff
@@ -1583,6 +1827,30 @@ def run_taxonomy_predictor(opt: TaxometerOptions):
...
 
 
+def run_taxonomy_cross_validation(opt: TaxometerOptions):
+    composition, abundance = load_composition_and_abundance(
+        opt.general,
+        opt.composition,
+        opt.abundance,
+        vamb.vambtools.BinSplitter.inert_splitter(),
+    )
+    abundance, tnfs, lengths, _ = (
```
Comment (jakobnissen) [inline vamb/__main__.py:1835]: Why assign the identifiers to `_`?

File: vamb/__main__.py [hunk 6]
```diff
@@ -1559,6 +1565,244 @@ def predict_taxonomy(
...
+    t = vamb.taxonomy.Taxonomy.from_file(tax.path, comp_metadata, False)
+    n_contigs = len(t.contig_taxonomies)
+    orig_path = str(tax.path)
+    predicted_path = out_dir.joinpath(f"results_taxonomy_predicted_kfold.tsv")
+    file_tracking = out_dir.joinpath(f"file_tracking.tsv")
+    accuracy_file = out_dir.joinpath(f"accuracy_report.tsv")
+
+    all_predicted_taxonomies = []
+    k = 5
+    kf = KFold(n_splits=k, shuffle=True)
```
Comment (jakobnissen) [inline vamb/__main__.py:1687]: Is it possible to seed the random shuffling using the seed from `GeneralOptions`? If so, please do it.

File: vamb/taxonomy.py [hunk 7]
```diff
@@ -176,6 +177,9 @@ def __init__(
         is_canonical: bool,
     ):
         if len(taxonomies) != len(metadata.identifiers):
+            logger.error(
+                f"Taxonomies length: {len(taxonomies)}, Identifiers length: {len(metadata.identifiers)}"
+            )
```
Comment (jakobnissen) [inline vamb/taxonomy.py:182]: I believe all exceptions are already caught and logged by the logger. Instead, improve the `ValueError`

File: vamb/taxonomy.py [hunk 8]
```diff
@@ -209,7 +213,18 @@ def parse_tax_file(
                 )
             for linenum_minus_two, line in enumerate(lines):
                 fields = line.split("\t")
-                if len(fields) != 3:
+                if len(fields) == 1:
```
Comment (jakobnissen) [inline vamb/taxonomy.py:216]: Good catch!

File: vamb/__main__.py [hunk 9]
```diff
@@ -1559,6 +1563,244 @@ def predict_taxonomy(
...
+
+def cross_validate_taxonomy(
+    comp_metadata: vamb.parsecontigs.CompositionMetaData,
+    abundance: np.ndarray,
+    tnfs: np.ndarray,
+    lengths: np.ndarray,
+    out_dir: Path,
+    taxonomy_options: TaxometerOptions,
+    cuda: bool,
+) -> vamb.taxonomy.PredictedTaxonomy:
```
Comment (jakobnissen) [inline vamb/__main__.py:1674]: This return type is incorrect

File: vamb/__main__.py [hunk 10]
```diff
@@ -1559,6 +1565,244 @@ def predict_taxonomy(
...
+def _read_tsv_to_dict(path, key_col="contigs", val_col="predictions"):
+    d = {}
+    with open(path, newline="") as f:
+        r = csv.DictReader(f, delimiter="\t")
+        for row in r:
+            d[row[key_col]] = row[val_col]
+    return d
+
+
+def _parse_taxonomy_string(s, max_levels):
```
[the remark this answers is not in the document]
Comment (sgalkina) [reply vamb/__main__.py:1575]: max_levels is more of a maximum level in the entire file

File: vamb/__main__.py (hunk 4 above)
Comment (sgalkina) [reply vamb/__main__.py:1787]: I checked, and it was faster. I had the same thoughts, but ended up speeding it up just for my own sake, and kept it like that. Less error prone if serial though, and easier to debug. I think you are right and I move it back to serial for the sake of stability

File: vamb/__main__.py (hunk 5 above)
Comment (sgalkina) [reply vamb/__main__.py:1835]: as an indication that I don't use them later

**Review by jakobnissen — APPROVED** [review]
Great addition - I added a small fix for type errors.
